File: sdks/ios/Sources/LegacyBaaSSDK/LegacyBaaSClient.swift
import Foundation
import Network
import Security

/// Main Legacy B2BaaS Platform Client for iOS
///
/// Provides access to all platform services including SWIFT processing,
/// blockchain routing, BaNCS integration, and analytics.
///
/// Example usage:
/// ```swift
/// let client = LegacyBaaSClient(apiKey: "your-api-key")
///
/// // Check platform health
/// client.getHealth { result in
///     switch result {
///     case .success(let health):
///         print("Platform status: \(health.status)")
///     case .failure(let error):
///         print("Error: \(error)")
///     }
/// }
///
/// // Process SWIFT MT103 message
/// let mt103 = MT103Message(
///     transactionReference: "TXN123456",
///     currency: "USD",
///     amount: 10000,
///     orderingCustomer: Customer(account: "123456789", name: "John Doe"),
///     beneficiaryCustomer: Customer(account: "987654321", name: "Jane Smith")
/// )
///
/// client.swift.processMT103(mt103) { result in
///     switch result {
///     case .success(let response):
///         print("Transaction ID: \(response.transactionId)")
///     case .failure(let error):
///         print("Error: \(error)")
///     }
/// }
/// ```
@objc public class LegacyBaaSClient: NSObject {
    
    // MARK: - Properties
    
    /// SDK version
    public static let sdkVersion = "1.0.0"
    
    /// Configuration options
    public let configuration: LegacyBaaSConfiguration
    
    /// SWIFT processing service
    public lazy var swift: SwiftService = SwiftService(client: self)
    
    /// Blockchain routing service
    public lazy var blockchain: BlockchainService = BlockchainService(client: self)
    
    /// BaNCS integration service
    public lazy var bancs: BancsService = BancsService(client: self)
    
    /// Analytics service
    public lazy var analytics: AnalyticsService = AnalyticsService(client: self)
    
    /// Webhook handler service
    public lazy var webhooks: WebhookService = WebhookService(client: self)
    
    private let session: URLSession
    private let apiKeyManager: APIKeyManager
    private let monitor: NWPathMonitor
    private let monitorQueue = DispatchQueue(label: "network-monitor")
    
    // MARK: - Initialization
    
    /// Initialize Legacy B2BaaS client
    /// - Parameters:
    ///   - apiKey: Your Legacy B2BaaS API key
    ///   - configuration: Optional configuration (uses default if not provided)
    public init(apiKey: String, configuration: LegacyBaaSConfiguration = .default) {
        guard APIKeyValidator.isValid(apiKey) else {
            fatalError("Invalid API key format")
        }
        
        self.configuration = configuration
        self.apiKeyManager = APIKeyManager(apiKey: apiKey)
        
        // Configure URL session
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = configuration.timeout
        config.timeoutIntervalForResource = configuration.timeout * 2
        config.requestCachePolicy = .reloadIgnoringLocalCacheData
        
        // Set default headers
        config.httpAdditionalHeaders = [
            "Accept": "application/json",
            "Content-Type": "application/json",
            "User-Agent": "LegacyBaaS-iOS-SDK/\(Self.sdkVersion)",
            "X-SDK-Version": Self.sdkVersion,
            "X-SDK-Language": "swift"
        ]
        
        self.session = URLSession(configuration: config)
        self.monitor = NWPathMonitor()
        
        super.init()
        
        // Start network monitoring
        monitor.start(queue: monitorQueue)
        
        if configuration.enableLogging {
            print("LegacyBaaS iOS SDK initialized for environment: \(configuration.environment)")
        }
    }
    
    deinit {
        monitor.cancel()
        session.invalidateAndCancel()
    }
    
    // MARK: - Public API
    
    /// Check platform health status
    /// - Parameter completion: Completion handler with health status result
    public func getHealth(completion: @escaping (Result<HealthStatus, LegacyBaaSError>) -> Void) {
        performRequest(endpoint: "/health", method: .GET, completion: completion)
    }
    
    /// Test API connectivity
    /// - Parameter completion: Completion handler with ping result
    public func ping(completion: @escaping (Result<PingResponse, LegacyBaaSError>) -> Void) {
        performRequest(endpoint: "/ping", method: .GET, completion: completion)
    }
    
    /// Get account information
    /// - Parameter completion: Completion handler with account details
    public func getAccount(completion: @escaping (Result<AccountInfo, LegacyBaaSError>) -> Void) {
        performRequest(endpoint: "/account", method: .GET, completion: completion)
    }
    
    /// Get API usage statistics
    /// - Parameters:
    ///   - period: Time period (day, week, month)
    ///   - completion: Completion handler with usage statistics
    public func getUsage(period: String = "day", completion: @escaping (Result<UsageInfo, LegacyBaaSError>) -> Void) {
        let endpoint = "/usage?period=\(period)"
        performRequest(endpoint: endpoint, method: .GET, completion: completion)
    }
    
    // MARK: - Internal Methods
    
    internal func performRequest<T: Codable>(
        endpoint: String,
        method: HTTPMethod,
        body: Codable? = nil,
        completion: @escaping (Result<T, LegacyBaaSError>) -> Void
    ) {
        guard let url = buildURL(endpoint: endpoint) else {
            completion(.failure(.invalidURL))
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method.rawValue
        
        // Add authorization header
        do {
            let apiKey = try apiKeyManager.getAPIKey()
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        } catch {
            completion(.failure(.authenticationError("Failed to retrieve API key")))
            return
        }
        
        // Add body if provided
        if let body = body {
            do {
                request.httpBody = try JSONEncoder().encode(body)
            } catch {
                completion(.failure(.encodingError(error)))
                return
            }
        }
        
        // Perform request with retry logic
        performRequestWithRetry(request: request, retryCount: 0, completion: completion)
    }
    
    private func performRequestWithRetry<T: Codable>(
        request: URLRequest,
        retryCount: Int,
        completion: @escaping (Result<T, LegacyBaaSError>) -> Void
    ) {
        session.dataTask(with: request) { [weak self] data, response, error in
            guard let self = self else { return }
            
            if let error = error {
                if retryCount < self.configuration.maxRetries && self.shouldRetry(error: error) {
                    DispatchQueue.global().asyncAfter(deadline: .now() + self.configuration.retryDelay) {
                        self.performRequestWithRetry(request: request, retryCount: retryCount + 1, completion: completion)
                    }
                    return
                }
                completion(.failure(.networkError(error)))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(.invalidResponse))
                return
            }
            
            guard let data = data else {
                completion(.failure(.noData))
                return
            }
            
            // Handle HTTP errors
            if let apiError = self.handleHTTPError(statusCode: httpResponse.statusCode, data: data) {
                if self.shouldRetry(statusCode: httpResponse.statusCode) && retryCount < self.configuration.maxRetries {
                    DispatchQueue.global().asyncAfter(deadline: .now() + self.configuration.retryDelay) {
                        self.performRequestWithRetry(request: request, retryCount: retryCount + 1, completion: completion)
                    }
                    return
                }
                completion(.failure(apiError))
                return
            }
            
            // Decode successful response
            do {
                let decoder = JSONDecoder()
                decoder.dateDecodingStrategy = .iso8601
                let result = try decoder.decode(T.self, from: data)
                completion(.success(result))
            } catch {
                completion(.failure(.decodingError(error)))
            }
        }.resume()
    }
    
    private func buildURL(endpoint: String) -> URL? {
        let baseURL = configuration.baseURL ?? configuration.environment.baseURL
        let fullURL = "\(baseURL)/api/v1\(endpoint)"
        return URL(string: fullURL)
    }
    
    private func shouldRetry(error: Error) -> Bool {
        if let urlError = error as? URLError {
            switch urlError.code {
            case .timedOut, .networkConnectionLost, .notConnectedToInternet:
                return true
            default:
                return false
            }
        }
        return false
    }
    
    private func shouldRetry(statusCode: Int) -> Bool {
        return [408, 429, 500, 502, 503, 504].contains(statusCode)
    }
    
    private func handleHTTPError(statusCode: Int, data: Data) -> LegacyBaaSError? {
        switch statusCode {
        case 200...299:
            return nil
        case 400:
            return .validationError(parseErrorMessage(data))
        case 401:
            return .authenticationError(parseErrorMessage(data))
        case 429:
            return .rateLimitError(parseErrorMessage(data))
        case 503:
            return .serviceUnavailableError(parseErrorMessage(data))
        default:
            return .httpError(statusCode, parseErrorMessage(data))
        }
    }
    
    private func parseErrorMessage(_ data: Data) -> String {
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let message = json["message"] as? String {
            return message
        }
        return "Unknown error"
    }
}

// MARK: - HTTP Method

internal enum HTTPMethod: String {
    case GET = "GET"
    case POST = "POST"
    case PUT = "PUT"
    case DELETE = "DELETE"
}

File: sdks/dotnet/LegacyBaaSClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using LegacyBaaS.SDK.Configuration;
using LegacyBaaS.SDK.Exceptions;
using LegacyBaaS.SDK.Models.Common;
using LegacyBaaS.SDK.Services;

namespace LegacyBaaS.SDK
{
    /// <summary>
    /// Main Legacy B2BaaS Platform Client for .NET
    /// 
    /// Provides access to all platform services including SWIFT processing,
    /// blockchain routing, BaNCS integration, and analytics.
    /// </summary>
    /// <example>
    /// <code>
    /// var client = new LegacyBaaSClient("your-api-key");
    /// 
    /// // Check platform health
    /// var health = await client.GetHealthAsync();
    /// Console.WriteLine($"Platform status: {health.Status}");
    /// 
    /// // Process SWIFT MT103 message
    /// var mt103 = new MT103Message
    /// {
    ///     TransactionReference = "TXN123456",
    ///     Currency = "USD",
    ///     Amount = 10000m,
    ///     OrderingCustomer = new Customer 
    ///     { 
    ///         Account = "123456789", 
    ///         Name = "John Doe" 
    ///     }
    /// };
    /// 
    /// var result = await client.Swift.ProcessMT103Async(mt103);
    /// Console.WriteLine($"Transaction ID: {result.TransactionId}");
    /// </code>
    /// </example>
    public class LegacyBaaSClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly LegacyBaaSOptions _options;
        private readonly ILogger<LegacyBaaSClient> _logger;
        private readonly JsonSerializerOptions _jsonOptions;
        private bool _disposed;

        /// <summary>
        /// Gets the SWIFT processing service
        /// </summary>
        public ISwiftService Swift { get; }

        /// <summary>
        /// Gets the blockchain routing service  
        /// </summary>
        public IBlockchainService Blockchain { get; }

        /// <summary>
        /// Gets the BaNCS integration service
        /// </summary>
        public IBancsService Bancs { get; }

        /// <summary>
        /// Gets the analytics service
        /// </summary>
        public IAnalyticsService Analytics { get; }

        /// <summary>
        /// Gets the webhook handler service
        /// </summary>
        public IWebhookService Webhooks { get; }

        /// <summary>
        /// Gets the SDK version
        /// </summary>
        public static string SdkVersion => "1.0.0";

        /// <summary>
        /// Initializes a new instance of the LegacyBaaSClient
        /// </summary>
        /// <param name="apiKey">Your Legacy B2BaaS API key</param>
        /// <param name="options">Optional configuration options</param>
        /// <param name="httpClient">Optional HTTP client (will create new if not provided)</param>
        /// <param name="logger">Optional logger instance</param>
        /// <exception cref="ArgumentException">Thrown when API key is null or empty</exception>
        /// <exception cref="AuthenticationException">Thrown when API key format is invalid</exception>
        public LegacyBaaSClient(
            string apiKey, 
            LegacyBaaSOptions? options = null,
            HttpClient? httpClient = null,
            ILogger<LegacyBaaSClient>? logger = null)
        {
            if (string.IsNullOrWhiteSpace(apiKey))
                throw new ArgumentException("API key cannot be null or empty", nameof(apiKey));

            if (!IsValidApiKey(apiKey))
                throw new AuthenticationException("Invalid API key format");

            _options = options ?? new LegacyBaaSOptions();
            _logger = logger ?? Microsoft.Extensions.Logging.Abstractions.NullLogger<LegacyBaaSClient>.Instance;

            // Configure JSON options
            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                WriteIndented = false,
                DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
            };

            // Configure HTTP client
            _httpClient = httpClient ?? CreateHttpClient(apiKey);

            // Initialize services
            Swift = new SwiftService(_httpClient, _options, _jsonOptions, _logger);
            Blockchain = new BlockchainService(_httpClient, _options, _jsonOptions, _logger);
            Bancs = new BancsService(_httpClient, _options, _jsonOptions, _logger);
            Analytics = new AnalyticsService(_httpClient, _options, _jsonOptions, _logger);
            Webhooks = new WebhookService(_httpClient, _options, _jsonOptions, _logger);

            _logger.LogInformation("Legacy B2BaaS Client initialized for environment: {Environment}", _options.Environment);
        }

        /// <summary>
        /// Initializes a new instance using IOptions pattern
        /// </summary>
        /// <param name="options">Configuration options</param>
        /// <param name="httpClient">HTTP client</param>
        /// <param name="logger">Logger instance</param>
        public LegacyBaaSClient(
            IOptions<LegacyBaaSOptions> options,
            HttpClient httpClient,
            ILogger<LegacyBaaSClient> logger)
            : this(options.Value.ApiKey, options.Value, httpClient, logger)
        {
        }

        /// <summary>
        /// Check platform health status
        /// </summary>
        /// <param name="cancellationToken">Cancellation token</param>
        /// <returns>Health status information</returns>
        public async Task<HealthStatus> GetHealthAsync(CancellationToken cancellationToken = default)
        {
            var response = await _httpClient.GetAsync("/health", cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var content = await response.Content.ReadAsStringAsync(cancellationToken);
            return JsonSerializer.Deserialize<HealthStatus>(content, _jsonOptions)!;
        }

        /// <summary>
        /// Test API connectivity
        /// </summary>
        /// <param name="cancellationToken">Cancellation token</param>
        /// <returns>Ping response with timestamp</returns>
        public async Task<Dictionary<string, object>> PingAsync(CancellationToken cancellationToken = default)
        {
            var response = await _httpClient.GetAsync("/ping", cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var content = await response.Content.ReadAsStringAsync(cancellationToken);
            return JsonSerializer.Deserialize<Dictionary<string, object>>(content, _jsonOptions)!;
        }

        /// <summary>
        /// Get account information
        /// </summary>
        /// <param name="cancellationToken">Cancellation token</param>
        /// <returns>Account details</returns>
        public async Task<Dictionary<string, object>> GetAccountAsync(CancellationToken cancellationToken = default)
        {
            var response = await _httpClient.GetAsync("/account", cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var content = await response.Content.ReadAsStringAsync(cancellationToken);
            return JsonSerializer.Deserialize<Dictionary<string, object>>(content, _jsonOptions)!;
        }

        /// <summary>
        /// Get API usage statistics
        /// </summary>
        /// <param name="period">Time period (day, week, month)</param>
        /// <param name="cancellationToken">Cancellation token</param>
        /// <returns>Usage statistics</returns>
        public async Task<Dictionary<string, object>> GetUsageAsync(
            string period = "day", 
            CancellationToken cancellationToken = default)
        {
            var response = await _httpClient.GetAsync($"/usage?period={period}", cancellationToken);
            response.EnsureSuccessStatusCode();
            
            var content = await response.Content.ReadAsStringAsync(cancellationToken);
            return JsonSerializer.Deserialize<Dictionary<string, object>>(content, _jsonOptions)!;
        }

        /// <summary>
        /// Get client configuration
        /// </summary>
        public LegacyBaaSOptions GetConfiguration() => _options;

        private HttpClient CreateHttpClient(string apiKey)
        {
            var httpClient = new HttpClient();
            
            httpClient.BaseAddress = new Uri(_options.BaseUrl);
            httpClient.Timeout = _options.Timeout;
            
            // Set default headers
            httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {apiKey}");
            httpClient.DefaultRequestHeaders.Add("Accept", "application/json");
            httpClient.DefaultRequestHeaders.Add("User-Agent", $"LegacyBaaS-DotNet-SDK/{SdkVersion}");
            httpClient.DefaultRequestHeaders.Add("X-SDK-Version", SdkVersion);
            httpClient.DefaultRequestHeaders.Add("X-SDK-Language", "csharp");

            return httpClient;
        }

        private static bool IsValidApiKey(string apiKey)
        {
            return !string.IsNullOrWhiteSpace(apiKey) && apiKey.Length >= 20;
        }

        /// <summary>
        /// Dispose resources
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Dispose resources
        /// </summary>
        /// <param name="disposing">Whether disposing</param>
        protected virtual void Dispose(bool disposing)
        {
            if (!_disposed && disposing)
            {
                _httpClient?.Dispose();
                _logger.LogInformation("Legacy B2BaaS Client disposed");
                _disposed = true;
            }
        }
    }
}